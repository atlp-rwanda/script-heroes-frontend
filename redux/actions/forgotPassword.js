import * as actions from "./types";
import http from '../../src/utils/axios';

export const forgotAction = (data) => (dispatch) =>{
 dispatch({ type: actions.SET_FORGOT_LOADING });
    http.post(
      "/api/forgotPassword",
      data
    )

    .then((res) => {
      dispatch({
        type: actions.SET_FORGOT_SUCCESS,
        payload: res.data.message,
      });
    })
    .catch((error) => {
      dispatch({
        type: actions.SET_FORGOT_ERROR,
        payload: error.response.data.err,
      });
    });
};
