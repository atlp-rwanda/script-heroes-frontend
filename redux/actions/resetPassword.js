import * as actions from "./types";
import http from '../../src/utils/axios';

export const resetAction = (data) => (dispatch) => {
    let token = data.token.token
    let password = data.password
    let confirmPassword = data.confirmPassword
    let resetData = {password, confirmPassword}
  dispatch({ type: actions.SET_RESET_LOADING });
  http
    .post(`/api/resetPassword/${token}`, resetData)
    .then((res) => {
      dispatch({
        type: actions.SET_RESET_SUCCESS,
        payload: res.data.message,
      });
    })
    .catch((error) => {
      dispatch({
        type: actions.SET_RESET_INVALID,
        payload: error.response.data.err,
      });
    });
};
