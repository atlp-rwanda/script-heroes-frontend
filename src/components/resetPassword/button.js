import React from 'react';
import {Button} from 'reactstrap';

const Mybutton = ({label})=>{
return(
    <Button type="submit">{label}</Button>
)
}
export default Mybutton;
