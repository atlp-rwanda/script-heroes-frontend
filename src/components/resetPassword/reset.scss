/* forgot password form */
.forgotPassword {
  min-height: 69vh;
  position: relative;
  margin: 30px;
  top: 100px;
}
#forgotForm {
  min-height: 300px;
  border: 3px solid #e9ecef;
  border-radius: 10px;
  background-color: #e9ecef;
  padding: 20px;
}
#forgotForm h1 {
  color: #728191;
  text-align: center;
  text-decoration: underline;
  font-family: sans-serif;
  font-weight: bold;
}
.forgotContainer Label {
  margin-top: 20px;
}
.forgotContainer input {
  margin-top: 20px;
}
.forgotContainer button {
  margin: 30px;
  // background-color: #544E45;
  color: white;
  text-align: center;
}
.forgotPassword img {
  max-width: 600px;
  max-height: 300px;
  margin: 20px;
}
#response {
  margin-top: 8px;
  color: #f0b106;
}
.success {
  min-height: 72vh;
}
#successForm {
  background-color: #e9ecef;
  position: relative;
  top: 200px;
  min-height: 200px;
}
#successForm h1 {
  position: relative;
  top: 25px;
}

/* reset password form */

.resetPassword {
  position: relative;
  top: 100px;
  min-height: 72vh;
}
#resetForm {
  min-height: 360px;
  max-width: 700px;
  border: 10px solid #e9ecef;
  border-radius: 10px;
  background-color: #e9ecef;
}
#resetForm h1 {
  margin: 10px;
  color: #728191;
  text-decoration: underline;
  font-weight: bolder;
}
.resetImg {
  max-width: 600px;
  max-height: 500px;
}
#resetBtn {
  margin: 10px;
}
.passwordInput {
  position: relative;
  display: flex;
}
.resetPassword i {
  position: absolute;
  top: 20%;
  right: 5%;
}
.resetPassword i:hover {
  color: rgb(236, 185, 16);
  cursor: pointer;
}

@media (max-width: 470px) {
  .forgotPassword {
    position: relative;
    top: 10px;
  }
  #forgotForm h1 {
    font-size: 18px;
  }
  .forgotPassword img {
    max-width: 300px;
    max-height: 200px;
  }
  .forgotContainer input::placeholder {
    font-size: 16px;
  }
}

@media (max-width: 470px) {
  .resetPassword {
    position: relative;
    top: 50px;
  }
  .resetImg {
    max-width: 400px;
    max-height: 400px;
    margin-bottom: 50px;
  }
  #resetForm h1 {
    font-size: 28px;
  }
  #resetForm Input::placeholder {
    font-size: 16px;
  }
}
