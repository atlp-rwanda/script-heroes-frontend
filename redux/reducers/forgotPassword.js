import * as actions from "../actions/types";

const initialState = {
  error: null,
  message: null,
  loading: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case actions.SET_FORGOT_LOADING:
      return {
        ...state,
        loading: true,
      };
    case actions.SET_FORGOT_ERROR:
      return {
        ...state,
        error: action.payload,
        loading: false,
      };
    case actions.SET_FORGOT_SUCCESS:
      return {
        ...state,
        message: action.payload,
        loading: false,
      };
    default:
      return state;
  }
};
