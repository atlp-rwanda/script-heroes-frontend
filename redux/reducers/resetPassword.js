import * as actions from "../actions/types";

const initialState = {
  error: null,
  message: null,
  err: null,
  loading: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case actions.SET_RESET_LOADING:
      return {
        ...state,
        loading: true,
      };
    case actions.SET_RESET_SUCCESS:
      return {
        ...state,
        message: action.payload,
        loading: false,
      };
    case actions.SET_RESET_INVALID:
      return {
        ...state,
        err: action.payload,
        loading: false,
      };
    default:
      return state;
  }
};
