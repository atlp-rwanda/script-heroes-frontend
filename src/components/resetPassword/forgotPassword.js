import React, { useState, useEffect } from "react";
import image from "../../../Assets/forgot.png";
import "./reset.scss";
import { useDispatch, useSelector } from "react-redux";
import { forgotAction } from "../../../redux/actions/forgotPassword";
import "bootstrap/dist/css/bootstrap.min.css";
import Mybutton from "./button";
import {
  Col,
  Button,
  Form,
  Label,
  Input,
  Spinner,
  Alert,
  Container,
  Row,
} from "reactstrap";

const ForgotPassword = () => {
  //Create states
  const dispatch = useDispatch();
  const [message, setMessage] = useState("");
  const [visible, setVisible] = useState(false);
  const [color, setColor] = useState("");
  const [email, setEmail] = useState("");

  //Add function to dispatch data when submitted and make input an object
  const handleSubmit = (e) => {
    const newEmail = { email };
    e.preventDefault();
    if (!email) setMessage(error);
    setEmail("");
    dispatch(forgotAction(newEmail));
  };

  //Get data from API
  const error = useSelector((state) => state.forgot.error);
  const response = useSelector((state) => state.forgot.message);
  const loading = useSelector((state) => state.forgot.loading);

  //Add changes incase a state changes
  useEffect(() => {
    if (error) {
      setMessage(error);
      setVisible("true");
      setColor("danger");
    }

    if (response) {
      setMessage(response);
      setVisible("true");
      setColor("success");
      setTimeout(() => {
        window.location = "./forgot/success";
      }, 3000);
    }
  }, [error, response]);

  return (
    <Container fluid>
      <div className="forgotPassword">
        <Row>
          <Col md="6">
            <Form id="forgotForm" onSubmit={handleSubmit}>
              <center>
                <h1>Forgot Password?</h1>
              </center>
              <div className="forgotContainer">
                <Label>Email</Label>
                <Input
                  id="emailinput"
                  value={email}
                  onChange={(e) => setEmail(e.target.value)}
                  placeholder="Enter your email"
                ></Input>
                <center>
                  <br></br>
                  <div>{loading && <Spinner color="primary" />}</div>
                  <Alert isOpen={visible} color={color}>
                    {message}
                  </Alert>
                  <Mybutton label="Submit"></Mybutton>
                </center>
              </div>
            </Form>
          </Col>
          <Col md="6">
            <img src={image} alt="image"></img>
          </Col>
        </Row>
      </div>
    </Container>
  );
};
const Success = () => {
  return (
    <Container>
      <div className="success">
        <center>
          <Form id="successForm">
            <center>
              <h1>
                Please Click On The Link Sent In Your Email To Reset Your
                Password .
              </h1>
            </center>
          </Form>
        </center>
      </div>
    </Container>
  );
};

export { ForgotPassword, Success };
