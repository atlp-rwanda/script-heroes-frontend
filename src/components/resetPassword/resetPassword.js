import React, { useState, useEffect } from "react";
import image from "../../../Assets/reset.png";
import "./reset.scss";
import { useDispatch, useSelector } from "react-redux";
import { resetAction } from "../../../redux/actions/resetPassword";
import {
  Col,
  Button,
  Form,
  Label,
  Input,
  Spinner,
  Alert,
  Container,
  Row,
} from "reactstrap";
import { useParams } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faEye } from "@fortawesome/free-solid-svg-icons";

const eye = <FontAwesomeIcon icon={faEye} />;

function ResetPassword() {
  //Create states
  const token = useParams();
  const dispatch = useDispatch();
  const [visibility2, setVisibility2] = useState("password");
  const [visibility, setVisibility] = useState("password");
  
  const togglePasswordVisiblity = () => {
    if (visibility == "password") {
      setVisibility("text");
    } else {
      setVisibility("password");
    }
  };
  const togglePasswordVisiblity2 = () => {
    if (visibility2 == "password") {
      setVisibility2("text");
    } else {
      setVisibility2("password");
    }
  };
  const [message, setMessage] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [visible, setVisible] = useState(false);
  const [color, setColor] = useState("");

  //Add function to dispatch data when submitted and make input an object
  const handleSubmit = (e) => {
    const newPassword = { password };
    const newConfirm = { confirmPassword };
    e.preventDefault();
    if (!(password && confirmPassword))  {
      setMessage(error);
    }
    if (password != confirmPassword) {
      setMessage('Passwords do not match')
      setVisible("true");
      setColor("danger");
    };
    dispatch(resetAction({ token, password, confirmPassword }));
  };

  //Get data from API
  const error = useSelector((state) => state.resetPassword.err);
  const response = useSelector((state) => state.resetPassword.message);
  const loading = useSelector((state) => state.resetPassword.loading);

  //Add changes incase a state changes
  useEffect(() => {
    if (error) {
      setMessage(error);
      setVisible("true");
      setColor("danger");
    }

    if (response) {
      setMessage(response);
      setVisible("true");
      setColor("success");
      setTimeout(() => {
        window.location = "../login";
      }, 3000);
    }
  }, [error, response]);

  return (
    <Container fluid>
      {/* <Navigation /> */}
      <div className="resetPassword">
        <Row>
          <Col md="6">
            <Form id="resetForm" onSubmit={handleSubmit}>
              <center>
                <h1>Reset Password</h1>
              </center>
              <div className="resetContainer">
                <Label>New Password</Label>
                <br></br>
                <div className="passwordInput">
                  <Input
                    type={visibility}
                    placeholder="Enter your new password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                  ></Input>
                  <i onClick={togglePasswordVisiblity}>{eye}</i>
                </div>
                <br></br>
                <label>Confirm Password</label>
                <br></br>
                <div className="passwordInput">
                  <Input
                    type={visibility2}
                    placeholder="Repeat your password"
                    value={confirmPassword}
                    onChange={(e) => setConfirmPassword(e.target.value)}
                  ></Input>
                  <i id = 'myEye' onClick={togglePasswordVisiblity2}>{eye}</i>
                </div>
                <center>
                  <div id="response">
                    {loading && <Spinner color="primary" />}
                  </div>
                  <Alert isOpen={visible} color={color}>
                    {message}
                  </Alert>
                  <Button type="submit" id="resetBtn">
                    Reset
                  </Button>
                </center>
              </div>
            </Form>
          </Col>
          <Col md="6">
            <img className="resetImg" src={image} alt="image"></img>
          </Col>
        </Row>
      </div>
      {/* <Footer /> */}
    </Container>
  );
}

export default ResetPassword;
