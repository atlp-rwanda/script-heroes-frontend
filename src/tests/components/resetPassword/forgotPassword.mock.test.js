import React from "react";
import { mount } from "enzyme";
import { ForgotPassword, Success } from "../../../components/resetPassword/forgotPassword";
import {useSelector, useDispatch} from 'react-redux'
import { act } from "react-dom/test-utils";

const mockDispatch = jest.fn();
const mockSelector = jest.fn();

jest.mock("react-redux", () => ({
  useSelector: () => mockSelector,
  useDispatch: () => mockDispatch,
}));

describe("forgot page", () => {
  let wrapper;
  beforeEach(() => {
    act(() => {
      wrapper = mount(<ForgotPassword />, <Success/>);
    });
  });
  it("Should render with no errors", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
