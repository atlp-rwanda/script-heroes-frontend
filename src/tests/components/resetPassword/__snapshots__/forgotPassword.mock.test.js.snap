// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forgot page Should render with no errors 1`] = `
<ForgotPassword>
  <Container
    fluid={true}
    tag="div"
  >
    <div
      className="container-fluid"
    >
      <div
        className="forgotPassword"
      >
        <Row
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <div
            className="row"
          >
            <Col
              md="6"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <div
                className="col-md-6"
              >
                <Form
                  id="forgotForm"
                  onSubmit={[Function]}
                  tag="form"
                >
                  <form
                    className=""
                    id="forgotForm"
                    onSubmit={[Function]}
                  >
                    <center>
                      <h1>
                        Forgot Password?
                      </h1>
                    </center>
                    <div
                      className="forgotContainer"
                    >
                      <Label
                        tag="label"
                        widths={
                          Array [
                            "xs",
                            "sm",
                            "md",
                            "lg",
                            "xl",
                          ]
                        }
                      >
                        <label
                          className=""
                        >
                          Email
                        </label>
                      </Label>
                      <Input
                        id="emailinput"
                        onChange={[Function]}
                        placeholder="Enter your email"
                        type="text"
                        value=""
                      >
                        <input
                          className="form-control"
                          id="emailinput"
                          onChange={[Function]}
                          placeholder="Enter your email"
                          type="text"
                          value=""
                        />
                      </Input>
                      <center>
                        <br />
                        <div>
                          <Spinner
                            color="primary"
                            tag="div"
                            type="border"
                          >
                            <div
                              className="spinner-border text-primary"
                              role="status"
                            >
                              <span
                                className="sr-only"
                              >
                                Loading...
                              </span>
                            </div>
                          </Spinner>
                        </div>
                        <Alert
                          closeAriaLabel="Close"
                          color="success"
                          fade={true}
                          isOpen="true"
                          tag="div"
                          transition={
                            Object {
                              "appear": true,
                              "baseClass": "fade",
                              "baseClassActive": "show",
                              "enter": true,
                              "exit": true,
                              "in": true,
                              "mountOnEnter": false,
                              "onEnter": [Function],
                              "onEntered": [Function],
                              "onEntering": [Function],
                              "onExit": [Function],
                              "onExited": [Function],
                              "onExiting": [Function],
                              "tag": "div",
                              "timeout": 150,
                              "unmountOnExit": true,
                            }
                          }
                        >
                          <Fade
                            appear={true}
                            baseClass="fade"
                            baseClassActive="show"
                            className="alert alert-success"
                            enter={true}
                            exit={true}
                            in="true"
                            mountOnEnter={false}
                            onEnter={[Function]}
                            onEntered={[Function]}
                            onEntering={[Function]}
                            onExit={[Function]}
                            onExited={[Function]}
                            onExiting={[Function]}
                            role="alert"
                            tag="div"
                            timeout={150}
                            unmountOnExit={true}
                          >
                            <Transition
                              appear={true}
                              enter={true}
                              exit={true}
                              in="true"
                              mountOnEnter={false}
                              onEnter={[Function]}
                              onEntered={[Function]}
                              onEntering={[Function]}
                              onExit={[Function]}
                              onExited={[Function]}
                              onExiting={[Function]}
                              timeout={150}
                              unmountOnExit={true}
                            >
                              <div
                                className="alert alert-success fade"
                                role="alert"
                              />
                            </Transition>
                          </Fade>
                        </Alert>
                        <Mybutton
                          label="Submit"
                        >
                          <Button
                            color="secondary"
                            tag="button"
                            type="submit"
                          >
                            <button
                              aria-label={null}
                              className="btn btn-secondary"
                              onClick={[Function]}
                              type="submit"
                            >
                              Submit
                            </button>
                          </Button>
                        </Mybutton>
                      </center>
                    </div>
                  </form>
                </Form>
              </div>
            </Col>
            <Col
              md="6"
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <div
                className="col-md-6"
              >
                <img
                  alt="image"
                  src="test-file-stub"
                />
              </div>
            </Col>
          </div>
        </Row>
      </div>
    </div>
  </Container>
</ForgotPassword>
`;
